### PR TITLE
add global clusterset

### PR DESCRIPTION
**Pull request by ldpliu**
Signed-off-by: ldpliu <daliu@redhat.com>
https://github.com/stolostron/backlog/issues/22994

### CODE REVIEW

**Comment by ldpliu** [comment]
/cc @qiujian16 @elgnay Please help to review it 
I am not sure if we should identify the `autoUpdateAnnotation` `cluster.open-cluster-management.io/autoupdate` for global set.
https://github.com/open-cluster-management-io/registration/pull/251/files#diff-fc2d0a93df947afba08cbd4f0db61e2abf4429ff9e20b48489a1f9776c1711bcR117

File: pkg/hub/managedclusterset/global_managedclusterset_controller.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+func (c *globalManagedClusterSetController) syncGlobalClusterSet(ctx context.Context, originalGlobalClusterSet *clusterv1beta1.ManagedClusterSet) error {
+	globalClusterSet := originalGlobalClusterSet.DeepCopy()
+
+	// if the annotation has set to disable, global clusterset controller will not work.
+	if hasAnnotation(globalClusterSet, autoUpdateAnnotation, "false") {
+		klog.V(4).Info("GlobalManagedClusterSetDisabled", "The GlobalManagedClusterSet is disabled by user")
+		return nil
+	}
+
+	// if globalClusterSet.Spec is changed, rollback the change by update it to the original value.
```
Comment (qiujian16) [inline pkg/hub/managedclusterset/global_managedclusterset_controller.go:111]: should we merge label/annotation here also?

File: pkg/hub/managedclusterset/global_managedclusterset_controller.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		return err
+	}
+
+	// if globalClusterSet is terminating, add the key to the controller queue with a second-long delay
+	if !globalClusterSet.DeletionTimestamp.IsZero() {
+		syncCtx.Queue().AddAfter(globalClusterSetName, 5*time.Second)
+		return nil
+	}
+
+	if err := c.syncGlobalClusterSet(ctx, globalClusterSet); err != nil {
```
Comment (qiujian16) [inline pkg/hub/managedclusterset/global_managedclusterset_controller.go]: I would call it applyGlobalClusterSet and merge labels/annotation in it.

File: pkg/hub/manager.go [hunk 3]
```diff
@@ -149,7 +149,11 @@ func RunControllerManager(ctx context.Context, controllerContext *controllercmd.
 			controllerContext.EventRecorder,
 		)
 	}
-
+	globalManagedClusterSetController := managedclusterset.NewGlobalManagedClusterSetController(
```
Comment (qiujian16) [inline pkg/hub/manager.go:152]: I think this should be in the same feature gate of defaultclusterset. And please also update the description of that feature gate.

File: pkg/hub/manager.go [hunk 4]
```diff
@@ -167,6 +171,7 @@ func RunControllerManager(ctx context.Context, controllerContext *controllercmd.
 	go addOnFeatureDiscoveryController.Run(ctx, 1)
 	if features.DefaultHubMutableFeatureGate.Enabled(features.DefaultClusterSet) {
```
Comment (qiujian16) [inline pkg/hub/manager.go:173]: I think we need to update the description of the feature gate also.

File: pkg/hub/managedclusterset/global_managedclusterset_controller.go (hunk 2 above)
Comment (ldpliu) [reply pkg/hub/managedclusterset/global_managedclusterset_controller.go]: Currently, we only update the global set spec.
And I also add a integration test case to handle it : https://github.com/open-cluster-management-io/registration/pull/251/files#diff-20b89e21483044eda78a48c9410652eacae6e58058adfb927b90dbece0af57a3R59

File: pkg/hub/manager.go (hunk 4 above)
Comment (ldpliu) [reply pkg/hub/manager.go:173]: done in https://github.com/open-cluster-management-io/registration/pull/251/files#diff-2f6213e2bd15bb256be4ed430e664c08612e86384445be797a20d5f56d48d101R32

File: pkg/hub/managedclusterset/global_managedclusterset_controller.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		WithSync(c.sync).
+		// use ResyncEvery to make sure:
+		// 1. create the global clusterset once controller is launched
+		// 2. the global clusterset be recreated once it is deleted for some reason
+		ResyncEvery(10*time.Second).
+		ToController("GlobalManagedClusterSetController", recorder)
+}
+
+func (c *globalManagedClusterSetController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	globalClusterSetName := GlobalManagedClusterSetName
```
Comment (qiujian16) [inline pkg/hub/managedclusterset/global_managedclusterset_controller.go]: this seems not necessary.

Comment (ldpliu) [reply pkg/hub/managedclusterset/global_managedclusterset_controller.go]: done

**Pull request merged**